### PR TITLE
[DTensor] Fix DTensor.from_local() returns DTensor with wrong size for uneven sharded tensor

### LINKED ISSUE #110762: [DTensor] DTensor.from_local() creates DTensor with wrong .shape/.size() for un-even sharding case
### 🐛 Describe the bug

DTensor has different int method, distribute_tensor() and DTensor.from_local(). When using DTensor.from_local(), the size of the global tensor seems wrong. Repro below in a test. When using from_local API, if we are sharding on dim 0, the size() of the DTensor would become (dim0_size * num_of_rank, dim_1_size). When we have an uneven sharding this would be wrong. For example, in the repro below, we create the following DTensor on each rank using from_local(), the size() of the DTensor should be [10, 2]. However, when we call .size(), we would get the following result. 

rank0: [3, 2] --> size(): [12, 2]
rank1: [3,  2] --> size(): [12, 2]
rank2: [3, 2] --> size(): [12, 2]
rank3: [1, 2] --> size(): [4, 2]

```
# Owner(s): ["oncall: distributed"]
import torch
import torch.distributed.checkpoint as dist_cp
from torch.distributed._shard.sharded_tensor import ShardedTensor

from torch.distributed._tensor import DTensor, init_device_mesh, Replicate, distribute_tensor, Shard
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
from torch.distributed.fsdp._shard_utils import _all_gather_sharded_tensor
from torch.distributed.fsdp.fully_sharded_data_parallel import StateDictType

from torch.distributed.tensor.parallel import PairwiseParallel, parallelize_module
from torch.testing._internal.common_utils import run_tests

from torch.testing._internal.distributed._tensor.common_dtensor import (
    DTensorTestBase,
    MLPModule,
    skip_if_lt_x_gpu,
    with_comms,
)
from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir


class MLPModule(torch.nn.Module):
    def __init__(self, device):
        super().__init__()
        torch.manual_seed(5)
        self.net1 = torch.nn.Linear(2, 10, device=device)

    def forward(self, x):
        return self.net1(x)

    def reset_parameters(self):
        self.net1.reset_parameters()


class DTensorSizeDebug(DTensorTestBase):
    @with_comms
    @skip_if_lt_x_gpu(2)
    @with_temp_dir
    def test_fsdp_to_tp(self):
        CHECKPOINT_DIR = self.temp_dir

        mesh_shape = (self.world_size,)
        device_mesh = init_device_mesh(self.device_type, mesh_shape)
        model = MLPModule(self.device_type).cuda(self.rank)
        # create a FSDP wrapped model
        fsdp_model = FSDP(model, use_orig_params=True, device_mesh=device_mesh)

        FSDP.set_state_dict_type(
            fsdp_model,
            StateDictType.SHARDED_STATE_DICT,
        )
        fsdp_state_dict = fsdp_model.state_dict()
        # the weight shape should be [10, 2] but it was [12, 2] for rank 0, rank1, and rank2
        # and [4, 2] for rank[3].
        # The size is wrong because the internal implmentation of dtensor state_dict is using DTensor.from_local API.
        # https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_shard_utils.py#L197
        for k,v in fsdp_state_dict.items():
            if "weight" in k:
                print(f"rank:{self.rank}, state_dict weight:{v}, weight.shape:{v.shape}\n")


        # the size of the dtensor is correct when we call distribute_tensor
        global_tensor = torch.randn(10, 2)
        dtensor = distribute_tensor(global_tensor, device_mesh, (Shard(0),))
        print(f"rank:{self.rank}, dtensor:{dtensor}, dtensor.shape:{dtensor.shape}")


        # seems the issue is that when we use the from_local API, then the size of DTensor is incorrect.
        if self.rank != 3:
            from_local_dtensor = DTensor.from_local(torch.randn(3, 2), device_mesh, (Shard(0),))
        else:
            from_local_dtensor = DTensor.from_local(torch.randn(1, 2), device_mesh, (Shard(0),))
        print(f"rank:{self.rank}, from_local_dtensor:{from_local_dtensor}, from_local_dtensor.shape:{from_local_dtensor.shape}")


if __name__ == "__main__":
    run_tests()
```

### Versions

Nightly

### PR DESCRIPTION
**Pull request by wz337**
Fixes #110762 

This PR:
fixes issue described in #110762 by adding kwarg for shape and stride when creating DTensor using `DTensor.from_local()`. When `shape` and `stride` are provided, we skip calcualtion for `tensor_shape` and `tensor_stride` using `compute_global_tensor_info()`, as `compute_global_tensor_info()` always assume even sharding.

Test plan:
```
python3 test/distributed/_tensor/test_dtensor.py -k test_from_local_uneven_sharding
python3 test/distributed/_tensor/test_dtensor.py -k test_from_local_uneven_sharding_raise_error
```

cc. @wanchaol

### CODE REVIEW

**Comment by wz337** [comment]
@wanchaol Seems I am missing some changes in backward? As it runs into error for `function _FromTorchTensorBackward returned an incorrect number of gradients (expected 6, got 4)`, although I am not entirely sure how this can affect the code path when shape and stride are not passed in to DTensor.from_local(). 

Could you help look at the test error to see if you have context on what I am missing? Thanks! 
https://github.com/pytorch/pytorch/actions/runs/6439226777/job/17486843208?pr=110781

File: torch/distributed/fsdp/_shard_utils.py [hunk 1]
```diff
@@ -191,4 +191,10 @@ def _create_chunk_dtensor(
     # HSDP placements: [Replicate(), Shard(0)]
     placements = [Replicate() for _ in range(device_mesh.ndim)]
     placements[-1] = shard_placement  # type: ignore[call-overload]
-    return DTensor.from_local(local_tensor, device_mesh, placements)
+    return DTensor.from_local(
```
Comment (wanchaol) [inline torch/distributed/fsdp/_shard_utils.py:194]: looking at the fsdp integration code, I think we can separate out two changes:
1. the first change is that I think we should avoid using non-public API from DTensor, specifically `shard_placement._split_tensor` is not public. The way FSDP should do is to not do manual split, but it should do `from_local(replicate()).redistribute(shard(0))`, this should also avoid the bug you seen, as in theory after summon_params, the params are replicated, we don't need to manually split them, DTensor can do this for us already, and the created DTensor should have the correct shape/offset
2. we should still fix `from_local` by passing shape/stride to that from_local can handle the uneven sharding case without communication. But this can be a PR on top of 1

**Review by wanchaol — COMMENTED** [review]
see my inlined comments. Let's fix things in two steps

**Comment by awgu** [comment]
@wz337  As a minor convenience, what do you think about allowing `mesh_shape` in `init_device_mesh()` to allow `-1` as a valid value in the shape?

A common pattern before for intra-node TP was `torch.arange(dist.get_world_size()).view(-1, torch.cuda.device_count())`. However, now we might do something like:
```
tp_size = torch.cuda.device_count()
global_world_size = dist.get_world_size()
assert global_world_size % tp_size == 0  # `.view()` would handle this assertion before
mesh_shape = (global_world_size // tp_size, tp_size)
```

**Comment by awgu** [comment]
@wz337 If not already planned, would it be possible to add a few functionalities to the parent device mesh APIs:
1. Support getting the global placements
    - If I have 1D submeshes for TP and DP each, I want to get the joint global placements with respect to the parent 2D mesh. For now, this is always `(dp_placement, tp_placement)` because (1) we do intra-node TP and (2) the parent device mesh APIs only support 1D TP and 1D DP. However, I think our implementation should handle the logic to use the right order based on which index the TP and DP submesh has in the parent mesh.
2. (Not high priority) Support 2D DP (i.e. HSDP) -- this exercises the indexing logic because in theory we could have split cases like `(DP replicate, TP, DP shard)` as opposed to contiguous ones like `(DP replicate, DP shard, TP)`. (Again, some organizations may not be useful right now, but I think our implementation should be able to handle it.)

**Comment by wz337** [comment]
> 1. Will `from_local(replicate()).redistribute(shard(0))` involve an all-gather? If so, what if I know the global shape and stride and want to avoid the all-gather? I hope we can support that.
> 2. What if I already have a DTensor sharded on a (sub)mesh, and I want to further shard it along a new (sub)mesh? (E.g., given tensor parallel DTensor, I want to shard with FSDP.) Is this supported?
> 
> (For 1, I am planning to use `DTensor` constructor directly.)

1. From my understand, from_local(replicate()) would not involve an all-gather. When placements is (Replicate(),), it broadcasts rank0 tensor to all rank behind the scenes. 
2. For 2, we currently do not support this yet, but this is something we are looking to build support for. 

@wanchaol Correct me if I am wrong on the answers.

**Comment by wz337** [comment]
> @wz337 As a minor convenience, what do you think about allowing `mesh_shape` in `init_device_mesh()` to allow `-1` as a valid value in the shape?
> 
> A common pattern before for intra-node TP was `torch.arange(dist.get_world_size()).view(-1, torch.cuda.device_count())`. However, now we might do something like:
> 
> ```
> tp_size = torch.cuda.device_count()
> global_world_size = dist.get_world_size()
> assert global_world_size % tp_size == 0  # `.view()` would handle this assertion before
> mesh_shape = (global_world_size // tp_size, tp_size)
> ```

Originally, we wanted to support this, but then we want init_device_mesh() to be a replacement for dist.init_process_group(). That means init_device_mesh() could be called without dist.init_process_group(), so the -1 would not work under this scenario. That is why it is currently not supported. 

@wanchaol Maybe we can add back support for this with additional check to see if dist initialization already happened beforer calling init_device_mesh().

**Comment by wz337** [comment]
> @wz337 If not already planned, would it be possible to add a few functionalities to the parent device mesh APIs:
> 
> 1. Support getting the global placements
>    
>    * If I have 1D submeshes for TP and DP each, I want to get the joint global placements with respect to the parent 2D mesh. For now, this is always `(dp_placement, tp_placement)` because (1) we do intra-node TP and (2) the parent device mesh APIs only support 1D TP and 1D DP. However, I think our implementation should handle the logic to use the right order based on which index the TP and DP submesh has in the parent mesh.
> 2. (Not high priority) Support 2D DP (i.e. HSDP) -- this exercises the indexing logic because in theory we could have split cases like `(DP replicate, TP, DP shard)` as opposed to contiguous ones like `(DP replicate, DP shard, TP)`. (Again, some organizations may not be useful right now, but I think our implementation should be able to handle it.)

For 1, we do have an API to get the index of child mesh in their parent mesh, which is get_parent_mesh_dim() in https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/device_mesh.py#L72. See usage in test case here: https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_device_mesh.py#L284 I think the bottleneck for allowing different order is the state_dict construction, which is tricky part, as our 2D FSDP + TP under today's implementation would always assume intra-node TP and inter-node FSDP.

2. Same as 1, I think we can support the indexing in device_mesh today, but we may need to investigate the state_dict() construction for different scenarios.

File: torch/distributed/fsdp/_shard_utils.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/fsdp/_shard_utils.py:194]: Yes, this makes sense. Thanks for the suggestions! @wanchaol 

As suggested, I separated out the changes in FSDP to use public API from DTensor here. https://github.com/pytorch/pytorch/pull/110831 

Will follow up with you regarding the fix for DTensor.from_local().

**Comment by wanchaol** [comment]
@awgu 
> 1. Will `from_local(replicate()).redistribute(shard(0))` involve an all-gather? If so, what if I know the global shape and stride and want to avoid the all-gather? I hope we can support that.

`from_local(replicate()).redistribute(shard(0))` won't invoke any collective and it would simply be local chunk, this is because the after summon_full_params in FSDP params already are replicated, we can create replicated DTensor first then transform it to sharded to reflect FSDP sharding layout, and since it's creating from replicated placement, DTensor know the correct shape/stride. if `from_local` from a `Shard` with uneven sharding, it is sth we still need to fix and passing in global shape/stride make sense.

> 2. What if I already have a DTensor sharded on a (sub)mesh, and I want to further shard it along a new (sub)mesh? (E.g., given tensor parallel DTensor, I want to shard with FSDP.) Is this supported?

This is sth we don't support today with `distribute_tensor` but it should be pretty easy to add.

File: torch/distributed/_tensor/api.py [hunk 2]
```diff
@@ -105,16 +105,19 @@ def forward(  # type: ignore[override]
...
-        tensor_shape, tensor_stride = compute_global_tensor_info(
-            input, device_mesh, placements
-        )
+        if not shape or not stride:
+            # if it's not by default run_check, we assume user is certain that each
+            # rank has the same tensor shape, and we just use that to calculate the
+            # global shape
+            tensor_shape, tensor_stride = compute_global_tensor_info(
+                input, device_mesh, placements
+            )
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:120]: can we just do:
```
else:
    tensor_shape=shape
    tensor_stride=stride
```
so that we don't need to check them again in the later callsites

File: torch/distributed/_tensor/api.py [hunk 3]
```diff
@@ -279,6 +282,8 @@ def from_local(
         placements: Optional[Sequence[Placement]] = None,
         *,
         run_check: bool = True,
+        shape: Optional[torch.Size] = None,
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:285]: can you add some docstr about these two fields?

File: torch/distributed/_tensor/api.py [hunk 4]
```diff
@@ -323,11 +328,19 @@ def from_local(
         if placements is None:
             placements = [Replicate() for _ in range(device_mesh.ndim)]
 
+        if (shape is not None) != (stride is not None):
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:331]: can we merge this check together to the autograd function above?

**Review by wanchaol — COMMENTED** [review]
looks good, have some comments about how the checks organized

File: torch/distributed/_tensor/api.py [hunk 5]
```diff
@@ -105,16 +105,23 @@ def forward(  # type: ignore[override]
...
         ctx.previous_placement = placements
         ctx.previous_device_mesh = device_mesh
 
-        # if it's not by default run_check, we assume user is certain that each
-        # rank has the same tensor shape, and we just use that to calculate the
-        # global shape
-        tensor_shape, tensor_stride = compute_global_tensor_info(
-            input, device_mesh, placements
-        )
+        if (shape is not None) != (stride is not None):
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:114]: hmmm this can be put to the below if condition? i.e. 
```
if shape and stride:
    tensor_shape, tensor_stride = shape, stride
elif shape is not None and stride is not None:
    ... compute global tensor info
else:
    raise RuntimeError

**Review by wanchaol — APPROVED** [review]
only have one nit, stamp to unblock, nice work!

File: torch/distributed/_tensor/api.py (hunk 5 above)
Comment (wz337) [reply torch/distributed/_tensor/api.py:114]: Oh ye. Good catch! Definitely not thinking there. lol
